docs/datatypes/tsemaphore.md

```diff
@@ -0,0 +1,99 @@
+---
+id: datatypes_tsemaphore
+title:  "TSemaphore"
+---
+`TSemaphore` is a semaphore with transactional semantics that can be used to control access to a common resource. It 
+holds a certain number of permits, and permits may be acquired or released.
+
+## Create a TSemaphore
+Creating a `TSemaphore` with 10 permits:
+```scala mdoc:silent
+import zio._
+import zio.stm._
+
+val tSemaphoreCreate: STM[Nothing, TSemaphore] = TSemaphore.make(10L)
+```
+
+## Acquire a permit
+Acquiring a permit reduces the number of remaining permits that the `TSemaphore` contains. Acquiring a permit is done 
+when a user wants to access a common shared resource:
+```scala mdoc:silent
+import zio._
+import zio.stm._
+
+val tSemaphoreAcq: UIO[TSemaphore] = (for {
+  tSem <- TSemaphore.make(2L)
+  _    <- tSem.acquire
+} yield tSem).commit
+```
+Note that if you try to acquire a permit when there are no more remaining permits in the semaphore then execution will be blocked until a permit is ready to be acquired.
```

Comment:
I know it is always the case but maybe emphasize that semantic blocking doesn't actually block any threads and won't loop again until one of the the number of permits changes.

Reply:
done :-)